src/api.rs

```diff
@@ -871,26 +890,63 @@ impl Client {
     pub async fn login(
         &self,
         email: &str,
+        sso_id: Option<&str>,
         device_id: &str,
         password_hash: &crate::locked::PasswordHash,
         two_factor_token: Option<&str>,
         two_factor_provider: Option<TwoFactorProviderType>,
     ) -> Result<(String, String, String)> {
-        let connect_req = ConnectPasswordReq {
-            grant_type: "password".to_string(),
-            username: email.to_string(),
-            password: Some(crate::base64::encode(password_hash.hash())),
-            scope: "api offline_access".to_string(),
-            client_id: "desktop".to_string(),
-            client_secret: None,
-            device_type: 8,
-            device_identifier: device_id.to_string(),
-            device_name: "rbw".to_string(),
-            device_push_token: String::new(),
-            two_factor_token: two_factor_token
-                .map(std::string::ToString::to_string),
-            two_factor_provider: two_factor_provider.map(|ty| ty as u32),
-        };
+        let connect_req;
+        match sso_id {
+            Some(sso_id) => {
+                let (sso_code, sso_code_verifier, callback_url) =
+                    self.obtain_sso_code(sso_id).await?;
+
+                connect_req = ConnectPasswordReq {
```

Comment:
I'd like to separate it into `ConnectPasswordReq` and `ConnectSSOReq`, both "inheriting" fields from `ConnectCommonReq`, not sure if (or rather how) same effect can be achieved with composition.

Reply:
something like https://serde.rs/attr-flatten.html would probably be what you want here

Reply:
Renamed to `ConnectTokenReq` so it corresponds better to the `/connect/token` that it's meant for and added subtypes for specific auth methods/strategies.